Enable enemy reveal when attacking friendly units in D2k

reaperrr opened the pull request:
We have FoW enabled by default, which the original didn't have, this somewhat helps.
The original also had this (like all WW 2D RTS games).

This also fixes #7621.

I think we should try to squeeze this into the next playtest/release.

reaperrr committed:
Enable RevealOnFire in D2k

reaperrr committed:
Remove bogus Buildable trait from D2k sandworm

reaperrr committed:
Make worms reveal shroud when swallowing friendly units

pchote commented:
I imagine this will be controversial, would you mind too much if we delayed this for Next + 1?

reaperrr commented:
> I imagine this will be controversial, would you mind too much if we delayed this for Next + 1?

I wouldn't really mind, I just think it wouldn't be remotely as controversial as for the other two official mods, but ok.

pchote commented:
If the general opinion is positive then I have no issue with taking it now, just wanted to be clear on the options.

ltem commented on mods/d2k/rules/defaults.yaml:
```diff
@@ -117,6 +117,9 @@
 		RequiresCondition: rank-elite
 		ZOffset: 256
 
+^RevealsShroudOnFire:
```
I accept that this works like it does but I would like to know why this trait is declared here and inherited for further use?

GraionDilach replied:
This allows the properties of the trait (like duration) being fine-tweaked only once.

ltem replied:
Thanks

ltem reviewed (commented):
- The worm fix is clever, I would have expected a much more difficult approach to solve that :)
- I don't have a strong opinion on `RevealOnFire` but I tend to support it.
- In the code itself I didn't found any issues and ingame it worked as expected but is there a reason why defensive structures are excluded?

Micr0Bit commented:
i will test this later

MustaphaTR commented:
I think this can be added to D2K mod and would actually be cool, but as long as it only reveals FoW and not shourd.

MustaphaTR commented:
Looks like this is added to all armed units and we have more armed units than unarmed. So why not add `Inherits@reveal: ^RevealsShroudOnFire` under `^Vehicle:`, `^Infantry` etc. and add `-RevealOnFire:` to unarmed units?

ltem commented:
I had some difficulties to observe the reveal, so as a little help it can be tested pretty good if you: 
- modify `mods/d2k/weapons/largeguns.yaml` range from `155mm` (Siege Tank Gun) from `Range: 5c512` to something ridiculous like `Range: 45c512` :comet: 
- play a skirmish game against the AI with `Debug Menu` enabled + `StartingUnits: Heavy Support` (enables possible Siege Tank)
- during the game enable `Disable Visibility Checks` and attack the enemy units with your (possible spawned) siege tank, surrender after some hits
- watch the replay from the AIs point of view to see the reveal in action (yeah it reveals shroud as well right now)

Bonus: Bogus action in the shellmap :D

pchote commented:
In cases like this it is best to run two clients in parallel in a MP game.

GraionDilach commented:
> as long as it only reveals FoW and not shourd.

Where's the logic in that? I don't think there's even support for such in the underlying code even.

MustaphaTR commented:
> Where's the logic in that? I don't think there's even support for such in the underlying code even.

Actually forget what i siad. For some reason i thought RevealOnFire wasn't a thing in original. Now i remember the holes Ornithopters /Carryalls create because of Missile Tanks.

reaperrr commented:
Updated.

- moved `RevealOnFire` onto actor defaults directly and then disabled it on unarmed actors. Exceptions: Sandworm (no inherited default) and Ornithopter (most aircraft are unarmed) implement it directly
- added `RevealOnFire` to defenses

ltem commented:
I think if this PR is merged, #7621 can be closed

GraionDilach reviewed (approved):
:+1:

reaperrr commented:
@OpenRA/engine-hackers This basically has +2 already and I'd like to see this make it into the playtest (we can still revert and delay the changes if feedback from pt is mostly negative), but I don't want to 'approve' my own PR just to make it mergeable, if possible.

pchote reviewed (approved):
I haven't tested this, but the idea is fine by me.

reaperrr commented:
[changelog](https://github.com/OpenRA/OpenRA/wiki/Changelog-(bleed)/_compare/46590e0d25d5fd74e22fed701b0fcd0b97fdf670...9a0bcbd91e233b63ed0a9d9ef2fc0f003329f45f)
